rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,206 @@
+2FA management:
+
+```sh
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+
+$ tsh 2fa add
+Adding a new 2FA device.
+Choose device type (1 - OTP, 2 - U2F): 2
+Enter device name: solokey
+Tap any *registered* security key... <tap>
+Tap your *new* security key... <tap>
+2FA device "solokey" added.
+
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+solokey            U2F    Wed 16 Dec 2020 02:05:46 PM PST
+
+$ tsh 2fa rm yubikey
+Tap any *registered* security key... <tap>
+2FA device "yubikey" removed.
+
+$ tsh 2fa rm "android OTP"
+Tap any *registered* security key... <tap>
+2FA device "android OTP" removed.
+
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+solokey            U2F    Wed 16 Dec 2020 02:06:46 PM PST
+
+# If 2FA is optional:
+$ tsh 2fa rm solokey
+You are about to remove the only remaining 2FA device.
+This will disable 2FA during login.
+Are you sure? (y/N): N
+
+# If 2FA is required:
+$ tsh 2fa rm solokey
+Can't remove the only remaining 2FA device.
+Please add a replacement 2FA device first using "tsh 2fa add".
+```
+
+#### web UI
+
+Web UI management of 2FA devices should be logically similar to the CLI:
+- a page to see all enrolled devices
+- a button to enroll a new device
+- buttons to remove any enrolled device (except for the last one)
+
+Web UI details, wireframes and implementation will be added later, when we have
+the capacity to do it. Initially, 2FA management is CLI-only.
+
+#### bootstrap
+
+Initially, a user account doesn't have a 2FA device. Depending on [cluster
+configuration](#configuration), use of 2FA devices might be required.
+
+If 2FA is required, a user is required to enroll a device during account
+creation (for local users) or first login (for SSO users).
+
+If 2FA is optional, a user can create an account and login without 2FA. They
+can then add 2FA devices as described above. If an existing user has at least
+one 2FA device registered, it's required during login.
+
+### configuration
+
+The current 2FA configuration in Teleport only applies to local users and is
+always required. We need to allow this to be optional (for users to migrate)
+and allow to mix different 2FA methods.
+
+Existing configuration options must be backwards-compatible - no change in
+behavior unless config values are changed.
+
+New fields added under `auth_service.authentication`:
+- `second_factor_optional` - bool, defaults to `false`; when `true`, users can
+  register and login without 2FA checks, and are able to add 2FA voluntarily
+- `second_factor` - keep existing values `otp`, `u2f`, `off`, and add a new
```

Comment:
I think `any` is great, can we remove extra knob `second_factor_optional`?

Reply:
`any` here doesn't include `off` - it just means that any 2FA method is supported
We could instead have other values:
- `second_factor: optional` - allow users to enroll in 2FA, but don't enforce it
- `second_factor: on` - require all users to have at least one 2FA method enabled

does that sound better?